backpack/core/derivatives/lstm.py

```diff
@@ -0,0 +1,358 @@
+            C_tanh_prod_t[:] = einsum(
+                "vnh, nh -> vnh",
+                C_prod[:, t],
+                1 - c_tanh[t] ** 2,
+            )
+
+            # product until node h
+            H_prod[:, t] = einsum(
+                "vnh, nh -> vnh",
+                IFGO_prod_t[:, :, H3:H4],
+                c_tanh[t],
+            ) + einsum(
+                "vnh, nh -> vnh",
+                C_tanh_prod_t,
+                ifgo[t, :, H3:H4],
+            )
+
+        return H_prod
+
+    def _jac_t_mat_prod(
+        self, module: LSTM, g_inp: Tuple[Tensor], g_out: Tuple[Tensor], mat: Tensor
+    ) -> Tensor:
+
+        self._check_parameters(module)
+
+        IFGO_prod: Tensor = self._ifgo_jac_t_mat_prod(module, mat)
+
+        X_prod: Tensor = einsum(
+            "vtnh, hi -> vtni",
+            IFGO_prod,
+            module.weight_ih_l0,
+        )
+        return X_prod
+
+    def _bias_ih_l0_jac_t_mat_prod(
+        self,
+        module: LSTM,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        self._check_parameters(module)
+
+        IFGO_prod: Tensor = self._ifgo_jac_t_mat_prod(module, mat)
+
+        if sum_batch:
+            eq = "vtnh -> vh"
+        else:
+            eq = "vtnh -> vnh"
+        return einsum(eq, IFGO_prod)
+
+    def _bias_hh_l0_jac_t_mat_prod(
+        self,
+        module: LSTM,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        self._check_parameters(module)
+
+        IFGO_prod: Tensor = self._ifgo_jac_t_mat_prod(module, mat)
+
+        if sum_batch:
+            eq = "vtnh -> vh"
+        else:
+            eq = "vtnh -> vnh"
+        return einsum(eq, IFGO_prod)
+
+    def _weight_ih_l0_jac_t_mat_prod(
+        self,
+        module: LSTM,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        self._check_parameters(module)
+
+        IFGO_prod: Tensor = self._ifgo_jac_t_mat_prod(module, mat)
+
+        if sum_batch:
+            eq = "vtnh, tni -> vhi"
+        else:
+            eq = "vtnh, tni -> vnhi"
+        return einsum(eq, IFGO_prod, module.input0)
+
+    def _weight_hh_l0_jac_t_mat_prod(
+        self,
+        module: LSTM,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        self._check_parameters(module)
+
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+
+        IFGO_prod: Tensor = self._ifgo_jac_t_mat_prod(module, mat)
+
+        if sum_batch:
+            eq = "vtnh, tng -> vhg"
+        else:
+            eq = "vtnh, tng -> vnhg"
+        return einsum(
+            eq,
+            IFGO_prod,
+            cat([zeros(1, N, H, device=mat.device), module.output[0:-1]], dim=0),
```

Comment:
It seems like the `einsum` multiplies `IFGO_prod` with a lot of zeros. To me it seems that this costs performance. Can this operation be rewritten?

Reply:
The zeros are actually only for the first time-step. So it would save 1/T of the operations. 
I could refactor this with an variable initialization and if-else statement. Still refactor?

Reply:
Fine. Just wanted to bring it up.